src/pmdas/ds389log/pmdads389log.1
'\"macro stdmacro
.\"
.\" Copyright (c) 2014-2018 Red Hat.
.\"
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; either version 2 of the License, or (at your
.\" option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
.\" or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
.\" for more details.
.\"
.\"
.TH PMDADS389LOG 1 "PCP" "Performance Co-Pilot"
.SH NAME
\f3pmdads389log\f1 \- 389 Directory Server Log PMDA
.SH DESCRIPTION
\fBpmdads389log\fP is a Performance Metrics Domain Agent (PMDA)
which extracts statistics from 389 Directory Server access log
using the \fBlogconv.pl\fP(1) utility.
.SH INSTALLATION
\fBpmdads389log\fP reads configuration file
.IP
.PD 0
.RS +4
.IP \(bu 2
.I \f(CR$PCP_PMDAS_DIR\fP/ds389log/ds389log.conf
.RE
.PD
.PP
This file can contain overridden values (Perl code) for the settings
listed at the start of
.IR pmdads389log.pl ,
namely:
.IP
.PD 0
.RS +4
.IP \(bu 2
\fBlogconv.pl\fR parameters
.IP \(bu
minimal query interval
.IP \(bu
389 DS access log to use
.IP \(bu
389 DS log directory to use (used if no file specified)
.IP \(bu
non-root user having read access to the access log file
.RE
.PD
.PP
Once this is setup, you can access the names and values for the
389 DS performance metrics by doing the following as root:
.sp 1
.RS +4
.ft B
.nf
# cd $PCP_PMDAS_DIR/ds389log
# ./Install
.fi
.ft P
.RE
.sp 1
To uninstall, do the following as root:
.sp 1
.RS +4
.ft B
.nf
# cd $PCP_PMDAS_DIR/ds389log
# ./Remove
.fi
.ft P
.RE
.sp 1
\fBpmdads389log\fP is launched by \fBpmcd\fP(1) and should never be
executed directly.
The Install and Remove scripts notify \fBpmcd\fP(1) when the
agent is installed or removed.
.SH FILES
.TP
.I \f(CR$PCP_PMDAS_DIR\fP/ds389log/ds389log.conf
configuration file for the \fBpmdads389log\fP agent
.TP
.I \f(CR$PCP_PMDAS_DIR\fP/ds389log/Install
installation script for the \fBpmdads389log\fP agent
.TP
.I \f(CR$PCP_PMDAS_DIR\fP/ds389log/Remove
undo installation script for the \fBpmdads389log\fP agent
.TP
.I \f(CR$PCP_LOG_DIR\fP/pmcd/ds389log.log
default log file for error messages from \fBpmdads389log\fP
.TP
.I /var/log/dirsrv/slapd\-.../access
389 Directory Server access log parsed by \fBlogconv.pl\fP
.SH PCP ENVIRONMENT
Environment variables with the prefix \fBPCP_\fP are used to parameterize
the file and directory names used by PCP.
On each installation, the
file \fI/etc/pcp.conf\fP contains the local values for these variables.
The \fB$PCP_CONF\fP variable may be used to specify an alternative
configuration file, as described in \fBpcp.conf\fP(5).
.SH SEE ALSO
.BR PCPIntro (1),
.BR logconv.pl (1),
.BR pmcd (1)
and
.BR pmdads389 (1).

.\" control lines for scripts/man-spell
.\" +ok+ pmdads {from pmdads389log} ds {from ds389log}
.\" +ok+ dirsrv slapd {both from /var/log/dirsrv/slapd} DS {from 389 DS ...}
.\" +ok+ logconv pl {both from logconv.pl}


CHANGELOG
pcp-6.3.6 (17 March 2025)
    - Debian build changes - see debian/changelog for details

pcp-6.3.4 (13 March 2025)
    - libpcp: fix endian issue in big-endian __pmTimestamp writing
    - pcp-xsos: fine-tune error handling on bad command line options
    - pcp-xsos: single-line pmPrintValue/pminfo values, escape shell chars
    - pcp-atop: correct BPF per-process network metric names
    - pmcd, pmdaproc: allow non-local username connection attributes
    - pmlogger, libpcp: fix handling of dynamic instance domain labels
    - pmrep: use BPF instead of BCC for proc networking metrics
    - pmlogconf: use libbpf over BCC for networking metrics
    - pmlogextract: fix logic around instances labels, allowing set updates
    - pmdalinux: mem.numa.hugepages metrics (per node, per hugepage size)
    - pmdabpftrace: fix for some python3 versions async IO create_task
    - pmdanvidia: if present use dso from the libnvidia-ml runtime package
    - pmdadm, pmieconf: default to enabling the dmthin pmie rules
    - configure.ac: fix autoreconf, add AC_DEFINE_UNQUOTED descriptions
    - rpm builds: enable bpf, bpftrace, bcc on riscv64 builds
    - rpm builds: drop user creation scriptlets for rpm builds
    - deb build: force debian/rules to be recreated each time
    - containers: updates to the archive-analysis container
    - selinux: several policy updates for new AVC denials
    - docs: improve SSL/TLS diagnostic in pcp-verify script

pcp-6.3.3 (29 January 2025)
    - Debian build changes - see debian/changelog for details

pcp-6.3.3 (29 January 2025)
    - Debian build changes - see debian/changelog for details

pcp-6.3.2 (6 November 2024)
    - PMDA additions, enhancements and bug fixes:
	pmdabpf: avoid a too-low rlimit setting within pmdabpf
	pmdaproc: add an access control mechanism for remote sampling
	pmdalinux: ensure divide-by-zero cannot happen with tmpfs.full
	pmdalinux: additional MPTcpExt netstat metrics
	pmdastatsd: keep correct track of pmns mapped metric count
	pmdaopenmetrics: run as pcp user rather than root by default
	pmdaopenmetrics: ensure correct cache file permissions used
    - Server-side utilities and log management scripts:
	libpcp, pmlogger: add support for zstd(1) compression of archives
	pmlogbasename: new, maps filename to PCP archive basename
	pmlogcompress: new, archive compression/decompression utility
	pmlogcheck: performance improvements
	pmlogcheck:  allow multiple archive options on command line
	pmlogextract: extend -v option with the full pmlogger syntax
	pmlogextract: fix for reporting file size on 32-bit platforms
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp: fix 64-bit big endian v3 archive timestamp decoding
	libpcp: change in archive value interpolation for non-counter metrics
	python api: allow timeval and timespec init with float/datetime
    - Client tools and utilities:
	pminfo: minor error path reworking
	pcp-buddyinfo: addition of a --no-interpol option
	pcp-ps: added option to view full task name with its argument
    - Build, infrastructure and packaging updates:
	build: deal with botched deb package names from Debian
	build: ensure pmie config.default created with correct ownership
	systemd: move pcp-reboot-init rpm spec logic from zeroconf to pcp
	selinux: ensure farm pmlogger/pmie processes labelled correctly
    - Documentation:
	docs: refine pmdaproc -A description details
	docs: make pmstat(1) match the code for -t defaults

pcp-6.3.1 (17 September 2024)
    - Client tools and utilities:
	pcp2arrow: fix extraction of values for metrics with indoms
	pcp-htop: fix platform setting enabling screen tabs
	pcp-htop: update to latest upstream bug fixes
	pcp-meminfo: added timestamp string for consistency
    - PMDA additions, enhancements and bug fixes:
	pmdaopenmetrics: add vLLM metrics support by default
	pmdahacluster: handle pacemaker 2.1.6 crm_mon output changes
	pmdahacluster: handle new style Pacemaker role descriptions
	pmdalinux: MPJoinSynBackupRx, MPJoinSynAckBackupRx netstat metrics
	pmdalinux: sysfs fallback from hinv.cpu.clock metric
	pmdaproc: lazy initialisation for accounting metrics
	pmdapmcd: expose pmcd limits as new metrics
	pmdammv: use isnan instead of open-coding NaN checks
    - Server-side utilities and log management scripts:
	pmcd: put an explicit cap on max number of fetched PMIDs
	pmcd: put an explicit cap on max number of contexts per client
	pmcd: cleanup caching of local hostname string
	pmpost: guard against possible symlink attack
	pmproxy: transition to dual Redis/Valkey support
    - libpcp, libpcp_pmda, libpcp_mmv, libpcp_web and language bindings:
	libpcp_web: fix a race in webgroup context teardown on timeout
	libpcp_web: transition to dual Redis/Valkey support
	libpcp: auto-load the proc PMDA in local context mode
	libpcp: fix buffer overflow in __pmServerAddInterface
	libpcp: refactor the attr PDU content checking
	libpcp: desc PDU consistency check and error message rework
	libpcp: text PDU consistency check and error message rework
	libpcp: creds PDU consistency check and error message rework
	libpcp: label PDU consistency check and error message rework
	libpcp: instance PDU consistency check and error message rework
	libpcp: rework some fetch PDU integrity checks and error messages
	libpcp: reworked __pmDecodeIDList integrity checks and warnings
	libpcp: ensure no fd leakage to child from __pmProcessPipe
	libpcp: ensure container names contain limited characters
	libpcp: fix buffer overrun test in __pmDecodeValueSet
	libpcp: hardening of the results PDU handling
	libpcp: hardening of the creds PDU handling
	libpcp: add new __pmCtlDebug() function
	python API: fix MMV metric text, indom and semantics handling
	python API: add log import timestamp convenience mechanisms
    - Build, infrastructure and packaging updates:
	build: tweak Makepkgs as dpkg-parsechangelog(1) is a bit picky
	build: drop gfs2 PMDA from build if distro vendor opts-out

src/pmdas/ds389log/pmdads389log.pl
#
# Copyright (C) 2014-2015 Marko Myllynen <myllynen@redhat.com>
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation; either version 2 of the License, or (at your
# option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
# for more details.
#

use strict;
use warnings;
use PCP::PMDA;
use Date::Manip;
use POSIX;

# 389 DS is not locale aware
setlocale(LC_ALL, "C");

our $lc_opts = '-D /dev/shm -s all';
our $lc_ival = 30; # minimal query interval in seconds, must be >= 30
our $ds_alog = ''; # empty - guess; ok if only one DS instance in use
our $ds_logd = '/var/log/dirsrv';
our $ds_user = 'dirsrv'; # empty - use root

# Configuration files for overriding the above settings
for my $file (pmda_config('PCP_PMDAS_DIR') . '/ds389log/ds389log.conf', './ds389log.conf') {
	eval `cat $file` unless ! -f $file;
}

my %data = (
	# logconv.pl string - name - subtree - cluster - id - type
	# type : 0 - cumulative, 1 - peak
	'Total Connections:'		=> [ 'totalconns', 'conns', 0, 0, 0 ],
	'Peak Concurrent Connections:'	=> [ 'peakconns', 'conns', 0, 1, 1 ],
	'U1'				=> [ 'cleanclose', 'conns', 0, 2, 0 ],
	'B1'				=> [ 'badclose', 'conns', 0, 3, 0 ],
	'Total Operations:'		=> [ 'totalops', 'ops', 1, 0, 0 ],
	'Total Results:'		=> [ 'totalres', 'ops', 1, 1, 0 ],
	'Searches:'			=> [ 'searches', 'ops', 1, 2, 0 ],
	'Modifications:'		=> [ 'mods', 'ops', 1, 3, 0 ],
	'Adds:'				=> [ 'adds', 'ops', 1, 4, 0 ],
	'Deletes:'			=> [ 'dels', 'ops', 1, 5, 0 ],
	'Mod RDNs:'			=> [ 'modrdns', 'ops', 1, 6, 0 ],
	'Compares:'			=> [ 'comps', 'ops', 1, 7, 0 ],
	'Binds:'			=> [ 'binds', 'ops', 1, 8, 0 ],
	'Paged Searches:'		=> [ 'pagedsearches', 'searches', 2, 0, 0 ],
	'Unindexed Searches:'		=> [ 'unindexedsearches', 'searches', 2, 1, 0 ],
	'err=0'				=> [ 'noerror', 'errors', 3, 0, 0 ],
	'err=X'				=> [ 'error', 'errors', 3, 1, 0 ], # custom
	'Highest FD Taken:'		=> [ 'fdhigh', 'fd', 4, 0, 1 ],
);

use vars qw( $pmda %metrics );

# Timestamps
my @lc_prev = localtime();
my @lc_curr;

sub ds389log_set_ds_access_log {
	$ds_alog = `ls -1 $ds_logd/slapd-*/access 2>/dev/null | tail -n 1`;
	my $un = `id -un`;
	chomp($ds_alog); chomp($un);
	die "$un can't read access log file \"$ds_logd/slapd-*/access\"" unless -f $ds_alog;
	$pmda->log("Using access log file $ds_alog");
}

sub ds389log_fetch {
	ds389log_set_ds_access_log() if $ds_alog eq '';
	return if $ds_alog eq '';

	# Server might not have written entries for operations during
	# the past few seconds yet so we will collect them next round.
	@lc_curr = localtime();
	$lc_curr[0] -= 30; # secs

	if ((strftime("%s", @lc_curr) - strftime("%s", @lc_prev)) < $lc_ival) {
		return;
	}

	# Don't include anything twice
	$lc_prev[0] += 1; # secs

	# Include the previous rotated log only if needed
	my $prev_log = `ls -1rt $ds_alog.2* 2>/dev/null | tail -n 1`;
	if ($prev_log ne '') {
		my $lastline = `tail -n 1 $prev_log`;
		$lastline =~ s/\[//;
		$lastline =~ s/\].*//;
		$lastline =~ s/:/ /;
		my $log_ts = UnixDate($lastline, "%s");
		if (strftime("%s", @lc_prev) > $log_ts) {
			$prev_log = '';
		}
	}

	my $lc_start = strftime("[%d/%b/%Y:%H:%M:%S %z]", @lc_prev);
	my $lc_end   = strftime("[%d/%b/%Y:%H:%M:%S %z]", @lc_curr);
	@lc_prev = @lc_curr;

	my $ds_stats = "logconv.pl -cpe $lc_opts -S $lc_start -E $lc_end $ds_alog $prev_log 2>/dev/null";
	open(STATS, "$ds_stats |") or
		die $pmda->err("pmda389log failed to open $ds_stats pipe: $!");
	my @stats = <STATS>;
	close(STATS);

	my $errors = 0; # combined
	my $nobind = 0; # no dupes
	foreach my $line (@stats) {
		my $key;

		if ($line =~ /^.*:/ || $line =~ /^U1/ || $line =~ /^B1/) {
			$key = $&;
		}

		if ($key eq "Binds:") {
			next if $nobind;
			$nobind = 1;
		}
		if ($line =~ /^err=.?/) {
			$key = 'err=X';
		}
		if ($line =~ /^err=0/) {
			$key = 'err=0';
		}

		if (defined($key) && defined $data{$key}) {
			$key = 'err=' if $key eq 'err=X';
			if ($line =~ /($key)\s+(\d+)/ || $line =~ /($key\d+)\s+(\d+)/) {
				my $value = $2;

				if ($key eq 'err=') {
					$key = 'err=X';
					$errors += $value;
					$value = $errors;
				}

				my $id = 'ds389log.' . $data{$key}[1] . '.' . $data{$key}[0];

				if ($data{$key}[4] eq 1) {
					my $prev = $metrics{$id};
					$value = $prev if $prev > $value;
				} else {
					$value = $metrics{$id} + $value;
				}

				$metrics{$id} = $value;
			}
		}
	}
}

sub ds389log_fetch_callback {
	my ($cluster, $item, $inst) = @_;

	if ($inst != PM_INDOM_NULL)	{ return (PM_ERR_INST, 0); }

	my $pmnm = pmda_pmid_name($cluster, $item);
	my $value = $metrics{$pmnm};

	if (!defined($value))		{ return (PM_ERR_APPVERSION, 0); }

	return ($value, 1);
}

$pmda = PCP::PMDA->new('ds389log', 131);

# Add and zero metrics
foreach my $key (keys %data) {
	my $name = 'ds389log.' . $data{$key}->[1] . '.' . $data{$key}->[0];
	$pmda->add_metric(pmda_pmid($data{$key}->[2], $data{$key}->[3]),
			PM_TYPE_U32, PM_INDOM_NULL, PM_SEM_COUNTER,
			pmda_units(0,0,1,0,0,PM_COUNT_ONE),
			$name, '', '');
	$metrics{$name} = 0;
}

$pmda->set_refresh(\&ds389log_fetch);
$pmda->set_fetch_callback(\&ds389log_fetch_callback);
# NB: needs to run as root or as a user having read access to the logs
$pmda->set_user($ds_user) if $ds_user ne '';
$pmda->run;
